mitodl/mit-open
# Change readable_id values for podcasts and episodes

### What are the relevant tickets?
Closes https://github.com/mitodl/hq/issues/4807
Closes https://github.com/mitodl/hq/issues/4834

### Description (What does it do?)
- Changes the `readable_id` value for podcasts and episodes to equal the guid value returned by the podcast RSS feeds.  Note: `readable_id` is a misleading name for the field, it should probably be `unique_id` instead because it's not necessarily a user-friendly id, but that's a potential issue for another day.
- Makes sure that all unpublished podcasts and episodes are removed from the search index (noticed this was missing when working on the above).
- Because this will unpublish all existing podcasts/episodes and create new ones with different ids, I included a new management command that will update any learning paths/userlists that include podcasts & episodes from the old unpublished resources to the new resources that match based on a specified field (in this case, should be url), and optionally delete the old unpublished resources afterward.  This could also be useful for any other types of resources that need their id's changed.  I might adopt this to another PR that is currently blocked for unrelated reasons (https://github.com/mitodl/mit-open/pull/1210)

```
# Usage: transfer_list_resources <resource_type> <matching_field> <from_source> <to_source> [--delete]
./manage.py transfer_list_resources podcast_episode url podcast podcast
./manage.py transfer_list_resources podcast url podcast podcast
```

### How can this be tested?
- On the main branch, if you don't already have podcasts imported, run `./manage.py backpopulate_podcast_data`
- Add some podcasts and episodes to learning paths and userlists.  Note which ones you add to which paths/lists so you can check on them later.
- Switch to this branch.  Run `./manage.py backpopulate_podcast_data` again, followed by these commands, with or without the delete option:
  ```
  ./manage.py transfer_list_resources podcast_episode url podcast podcast [--delete]
  ./manage.py transfer_list_resources podcast url podcast podcast [--delete]
  ```
- Check the relevant paths/lists and make sure they include the podcasts/episodes you added before, and that they now point to new resources with the correct `readable_id` values.
- Search for podcasts/episodes by title, you should not get any dupes in results.
